lib/setMode.js

```diff
@@ -1,21 +1,160 @@
+  } else if (typeof data === 'number') {
+    if(data < 0 || data >= constants.colors.length) throw 'invalid color for ' + region + ' region';
+    regionDef = {primary: {color: data, level: 0}, secondary: {color: constants.colors['off'], level: 0}};
+  } else if (typeof data === 'object') {
+    if (data.primary) {
+      regionDef = {primary: getColorAndIntensityLevel(data.primary, region, true)};
+    } else {
+      regionDef = {primary: getColorAndIntensityLevel(data, region, true)};
+    }
+    if (data.secondary) {
+      regionDef.secondary = getColorAndIntensityLevel(data.secondary, region);
+      //Default to off if both settings are unset
+      if (regionDef.secondary.color < 0 && regionDef.secondary.level < 0) regionDef.secondary = {color: constants.colors['off'], level: 0};
+      //Default to primary settings if either are unset
+      if (regionDef.secondary.color < 0) regionDef.secondary.color = regionDef.primary.color;
+      if (regionDef.secondary.level < 0) regionDef.secondary.level = regionDef.primary.level;
+    } else {
+      //Default secondary to off if unset
+      regionDef.secondary = {color: constants.colors['off'], level: 0};
+    }
+  } else 'invalid color for ' + region + ' region';
+  
+  return regionDef;
+}
+
+function getColorAndIntensityLevel(data, region, isRequired) {
+  var colorName, levelName;
+  var colorId = -1, levelId = -1;
+  
+  if (typeof data === 'string') {
+    colorName = data;
+  } else if (typeof data === 'number') {
+    colorId = data;
+  } else if (data.color) {
+    if (typeof data.color === 'string') {
+      colorName = data.color;
+    } else if (typeof data.color === 'number') {
+      colorId = data.color;
+    }
+  }
+  
+  if (colorName) {
+    if (typeof constants.colors[colorName] === 'undefined') throw 'invalid color for ' + region + ' region';
+    colorId = constants.colors[colorName];
+  }
+  
+  //Always invalid if color is above range. colorId < 0 should be allowed if the fields are not required
+  if(colorId >= constants.colors.length || (isRequired && colorId < 0)) throw 'invalid color for ' + region + ' region';
+    
+  
+  if (data.intensity) {
+    if (typeof data.intensity === 'string') {
+      levelName = data.intensity;
+    } else if (typeof data.intensity === 'number') {
+      levelId = data.intensity;
+    }
+  }
+  
+  if (levelName) {
+    if (typeof constants.levels[levelName] === 'undefined') throw 'invalid intensity for ' + region + ' region';
+    levelId = constants.levels[levelName];
+  }
+
+  if (levelId >= constants.colors.length) throw 'invalid intensity for ' + region + ' region';
+    
+  if (isRequired && levelId < 0) {
+    levelId = 0;
+  }
+
+  colorDef = {color: colorId, level: levelId};
+  
+  return colorDef;
+}
+
+function sendData(keyboard, val1, val2, val3, val4) {
+    var data = new Buffer(8);
+
+    data[0] = 0x01;
```

Comment:
Is there a reason why you changed a 1 to a 1 in hexadecimal notation?
0x01 in JavaScript is converted to a 1.